pkg/server/api/middleware/authorization_opa.go

```diff
@@ -0,0 +1,130 @@
+package middleware
+
+import (
+	"context"
+	"errors"
+
+	"github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+	"github.com/spiffe/spire/pkg/server/api/rpccontext"
+	"github.com/spiffe/spire/pkg/server/policy"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func (m *authorizationMiddleware) opaAuth(ctx context.Context, req interface{}, fullMethod string) (context.Context, bool, error) {
+	if m.policyEngine == nil {
+		return ctx, false, errors.New("No policy engine object found")
+	}
+
+	// Get SPIFFE ID
+	var spiffeID string
+	id, ok := rpccontext.CallerID(ctx)
+	if ok {
+		spiffeID = id.String()
+	}
+
+	input := policy.Input{
+		Caller:     spiffeID,
+		FullMethod: fullMethod,
+		Req:        req,
+	}
+
+	result, err := m.policyEngine.Eval(ctx, input)
+	if err != nil {
+		return ctx, false, err
+	}
+
+	ctx, allow, err := reconcileResult(ctx, result, m.entryFetcher, m.agentAuthorizer)
+	if err != nil {
+		return ctx, false, err
+	}
+
+	return ctx, allow, nil
+}
+
+func reconcileResult(ctx context.Context, res policy.Result, entryFetcher EntryFetcher, agentAuthorizer AgentAuthorizer) (context.Context, bool, error) {
+	// Check things in order of cost
+	if res.Allow {
+		return ctx, true, nil
+	}
+
+	// Check local
+	if res.AllowIfLocal && rpccontext.CallerIsLocal(ctx) {
+		return ctx, true, nil
+	}
+
+	// Get entries
+	if res.AllowIfAdmin || res.AllowIfDownstream {
+		ctx, entries, err := WithCallerEntries(ctx, entryFetcher)
+		if err != nil {
+			return ctx, false, err
+		}
+
+		if res.AllowIfAdmin {
+			if ctx, ok := isAdmin(ctx, entries); ok {
+				return ctx, true, nil
+			}
+		}
+
+		if res.AllowIfDownstream {
+			if ctx, ok := isDownstream(ctx, entries); ok {
+				return ctx, true, nil
+			}
+		}
+	}
+
+	if res.AllowIfAgent {
+		ctx, isAgent, err := isAgent(ctx, agentAuthorizer)
+		if err == nil && isAgent {
```

Comment:
Yea good catch.

Just to provide some context for when I was writing this.. 

`isAgent` may return an error if it can't figure out if its an agent, but if an authorization policy says please allow "if agent" or "if local" then this call to `isAgent` would error.. so the order of operations is what really "saves" it here. 

Maybe... the way to address this is changing `isAgent` such that it will not return an error if caller does not have an `x509`?